Partial Messages Extension

MarcoPolo opened the pull request:
Implements https://github.com/libp2p/specs/pull/685

Commits are meaningful. Best to review commit by commit.

MarcoPolo committed:
update protobufs for partial messages

MarcoPolo committed:
add structured rpc logging

MarcoPolo committed:
implement Partial Messages

MarcoPolo committed:
add partial messages to gossipsub router

haxxpop commented on extensions.go:
```diff
@@ -50,6 +56,14 @@ func (pe *PeerExtensions) ExtendRPC(rpc *RPC) *RPC {
+		if rpc.Control == nil {
+			rpc.Control = &pubsub_pb.ControlMessage{}
+		}
+		rpc.Control.Extensions = &pubsub_pb.ControlExtensions{
```
I think this removes the previously added `TestExtension` field above because this creates a new instance of `ControlExtensions`.

MarcoPolo replied:
yes, thank you. It was an incorrect conflict resolution

haxxpop commented on partialmessages/partialmsgs.go:
```diff
@@ -0,0 +1,395 @@
+	//   - Is the groupID well formed per application semantics?
+	//   - If this is a PartialIHAVE/PartialIWant, is the request metadata within
+	//     expected bounds?
+	ValidateRPC func(from peer.ID, rpc *pb.PartialMessagesExtension) error
```
This can be nil and this file assumes that this is not nil. The problem is that the users can create an instance of this struct with this to be nil and don't know that it's wrong.

I think we should have a constructor for this struct (`NewPartialMessageExtension`) and emit an error if this is nil or just assign this to a default value.

MarcoPolo replied:
This shouldn't be nil. It would be a user error to set this to nil and be detected as soon as you receive the first message.

However, an explicit error is nice. I don't want a needless constructor when we don't need one. Instead I'll add a .Validate method, and have that called in `WithPartialMessagesExtension`

haxxpop commented on partialmessages/partialmsgs.go:
```diff
@@ -0,0 +1,392 @@
+	}
+	state, ok := statePerTopic[string(groupID)]
+	if !ok {
+		state = newPartialMessageStatePerTopicGroup(max(e.GroupTTLByHeatbeat, minGroupTTL))
```
max is already done in `newPartialMessageStatePerTopicGroup`
```suggestion
		state = newPartialMessageStatePerTopicGroup(e.GroupTTLByHeatbeat)
```

haxxpop commented on partialmessages/partialmsgs.go:
```diff
@@ -0,0 +1,392 @@
+			inResponseToIWant = true
+		}
+
+		if !inResponseToIWant && eagerPush != nil && !bytes.Equal(ihave, peerState.has) {
```
I don't quite understand why do you need this condition? `!bytes.Equal(ihave, peerState.has) `. `eagerPush` is not part of the partial message, right?

MarcoPolo replied:
added a comment to explain

haxxpop commented on partialmessages/partialmsgs.go:
```diff
@@ -0,0 +1,392 @@
+	has      []byte
+	sentWant []byte
+	sentHas  []byte
+}
```
Could you please explain all the fields as comments here? I think it would be helpful while I read the code.

haxxpop commented on partialmessages/partialmsgs.go:
```diff
@@ -0,0 +1,392 @@
+		// Only send IHAVE if they don't have what we have and they haven't sent
+		// us an IDONTWANT
+		if !inResponseToIWant && partialIHAVE != nil {
+			if !peerStateOk || !bytes.Equal(ihave, peerState.sentHas) {
```
As commented here
```
		// Only send IHAVE if they don't have what we have and they haven't sent
		// us an IDONTWANT
```
I think this uses a wrong field. Can you confirm?
```suggestion
			if !peerStateOk || !bytes.Equal(ihave, peerState.has) {
```

MarcoPolo replied:
The comment was wrong. Updated.

haxxpop commented on partialmessages/partialmsgs.go:
```diff
@@ -0,0 +1,392 @@
+			}
+		}
+
+		// Only send an IWANT if it was different then before and we haven't exceeded eager limit
```
"and we haven't exceeded eager limit". I can't see where is the eager limit. Can you point it out please?

MarcoPolo replied:
stale comment, thanks

MarcoPolo commented:
thanks for the review @ppopth, friendly reminder you can see the latest changes with `git range-diff master 7fc7eb495bd6 578e12f4afcd`

haxxpop reviewed (approved):
Already looking good to me

sukunrt commented on pubsub.go:
```diff
@@ -262,6 +262,35 @@ type RPC struct {
+// LogValue implements slog.LogValuer.
+func (rpc *RPC) LogValue() slog.Value {
+	// Messages
+	var msgs []any
```
Can prealloc to len(rpc.Publish)

sukunrt commented on pubsub.go:
```diff
@@ -262,6 +262,35 @@ type RPC struct {
+		))
+	}
+
+	var fields []slog.Attr
```
Can prealloc.

sukunrt commented on extensions.go:
```diff
@@ -132,14 +149,79 @@ func (es *extensionsState) extensionsAddPeer(id peer.ID) {
+
+type routerForPartialMessage struct {
+	gs *GossipSubRouter
+}
```
NIT: I prefer `PartialMessageRouter`

sukunrt commented on partialmessages/partialmsgs.go:
```diff
@@ -0,0 +1,402 @@
+		iwant []byte,
+		ihave []byte,
+		partialMessageBytes []byte,
+	)
```
Should we just make it `OnIncomingRPC(from, rpc)`. The user anyway needs to deal with the RPC because of `ValidateRPC`. Also seems like any changes to the RPC will be bubbled up here.

sukunrt commented on partialmessages/partialmsgs.go:
```diff
@@ -0,0 +1,402 @@
+			state.peerState[p] = peerState
+		}
+		// Try to fulfill any outstanding PartialIWants
+		if peerStateOk && len(peerState.wants) > 0 {
```
NIT: `peerStateOk` check is redundant

sukunrt commented on partialmessages/partialmsgs.go:
```diff
@@ -0,0 +1,402 @@
+				if e.Logger != nil {
+					e.Logger.Warn("partial message extension failed to get partial message bytes", "error", err)
+				}
+				// Possibly a bad request, we'll delete the request as we will likely error next time we try to handle it
```
Should we log this deletion? A hexdump of the metadata maybe?

sukunrt commented on partialmessages/partialmsgs.go:
```diff
@@ -0,0 +1,402 @@
+			sendRPC = true
+			log.Debug("Eager pushing")
+			rpc.Message = eagerPush
+		}
```
Should this be another callback on PartialMessage? the bytes.Equal check isn't sufficient, we need a subset check. You have part 1 and peer has parts 1, 2, 3, 4. You'll still eager push, right?

sukunrt commented on partialmessages/partialmsgs.go:
```diff
@@ -0,0 +1,402 @@
+	//
+	// If the Partial Message is empty, the implementation MUST return:
+	// nil, metadata, nil.
+	PartialMessageBytesFromMetadata(metadata []byte) ([]byte, []byte, error)
```
I think we should just name this IWant: Makes it clearer. 
```
PartialMessageBytesFromIWantRepr(iwantrepr []byte) (partialMsgs []byte, newIWantRepr []byte, err error)
```

sukunrt commented on partialmessages/partialmsgs.go:
```diff
@@ -0,0 +1,402 @@
+
+	// AvailableParts returns the opaque metadata representing the parts this
+	// message contains. Used for PartialIHAVE
+	AvailableParts() ([]byte, error)
```
```suggestion
	// AvailableParts returns the opaque metadata representing the parts this
	// message contains. Used for PartialIHAVE
	AvailableParts() (iHaveRepr []byte, error)
```

sukunrt commented on partialmessages/partialmsgs.go:
```diff
@@ -0,0 +1,402 @@
+	// What parts the peer has
+	has []byte
+	// The wants this endpoint has sent to the peer
+	sentWant []byte
```
```suggestion
	// The wants we have sent to the peer
	sentWant []byte
```

sukunrt commented on partialmessages/partialmsgs.go:
```diff
@@ -0,0 +1,402 @@
+	// The wants this endpoint has sent to the peer
+	sentWant []byte
+	// The has this endpoint has sent to the peer
+	sentHas []byte
```
```suggestion
	// The has we have sent to the peer
	sentHas []byte
```

sukunrt commented on partialmessages/partialmsgs.go:
```diff
@@ -0,0 +1,402 @@
+	// MissingParts returns the opaque request metadata representing the missing
+	// parts. Used for PartialIWANT
+	// Return an empty slice if you have all available parts.
+	MissingParts() ([]byte, error)
```
```suggestion
	// MissingParts returns the opaque request metadata representing the missing
	// parts. Used for PartialIWANT
	// Return an empty slice if you have all available parts.
	MissingParts() (iWantRepr []byte, err error)
```

MarcoPolo committed:
Add documentation for the RequestPartialMessages topic option

and fix a bug where we would skip sending them full messages if they
requested partial messages and we disabled partial messages.

sukunrt commented on gossipsub_test.go:
```diff
@@ -4394,3 +4397,396 @@ func TestTestExtension(t *testing.T) {
+
+func TestPartialMessages(t *testing.T) {
+	topic := "test-topic"
+	const hostCount = 2
```
Should we make this a bit higher, say 5?

sukunrt commented on partialmessages/partialmsgs.go:
```diff
@@ -0,0 +1,290 @@
+	//
+	// If the Partial Message is empty, the implementation MUST return:
+	// nil, metadata, nil.
+	PartialMessageBytes(metadata []byte) ([]byte, []byte, error)
```
```suggestion
	PartialMessageBytes(hasPartsMetadata []byte) (msg []byte, newHasPartsMetadata []byte, _ error)
```

MarcoPolo replied:
I'll make this change. Though it's just partsMetadata as the meaning is left to the application.

sukunrt commented on partialmessages/partialmsgs.go:
```diff
@@ -0,0 +1,290 @@
+// keeps a reference to this object, so the implementation should either not
+// mutate this object in a separate goroutine after handing it to Gossipsub, or
+// take care to make the object thread safe.
+type PartialMessage interface {
```
NIT: partialmessages is the package name, so we can remove Partial from the type names:

read by users as `partialmessages.Message`

```suggestion
type Message interface {
```

sukunrt commented on partialmessages/partialmsgs.go:
```diff
@@ -0,0 +1,290 @@
+
+// partialMessagePeerState is the state we keep per peer. Used to make Publish
+// Idempotent.
+type partialMessagePeerState struct {
```
NIT: for the same reasons as the renaming for PartialMessage. 
```suggestion
type peerState struct {
```

sukunrt commented on partialmessages/partialmsgs.go:
```diff
@@ -0,0 +1,290 @@
+// Idempotent.
+type partialMessagePeerState struct {
+	// The parts metadata the peer has sent us
+	partsMetadata []byte
```
Is this the parts the peer has? If so
partsMetadata -> hasMetadata.

MarcoPolo replied:
Not necessarily. It can be used that way, but it could also be the inverse, a bloom filter or something else.

sukunrt commented on partialmessages/partialmsgs.go:
```diff
@@ -0,0 +1,290 @@
+	// If you need to do slow work, dispatch the work to your own goroutine.
+	OnIncomingRPC func(from peer.ID, rpc *pb.PartialMessagesExtension) error
+
+	// ValidateRequestMetadata should be a fast function that performs some
```
```suggestion
	// ValidateRPC should be a fast function that performs some
```

sukunrt commented on partialmessages/partialmsgs.go:
```diff
@@ -0,0 +1,290 @@
+func (e *PartialMessageExtension) groupState(topic string, groupID []byte) (*partialMessageStatePerTopicGroup, error) {
+	if e.statePerTopicPerGroup == nil {
+		e.statePerTopicPerGroup = make(map[string]map[string]*partialMessageStatePerTopicGroup)
+	}
```
NIT: we can initialize this in Init

sukunrt commented on partialmessages/partialmsgs.go:
```diff
@@ -0,0 +1,290 @@
+
+type Router interface {
+	SendRPC(p peer.ID, r *pb.PartialMessagesExtension, urgent bool)
+	MeshPeers(topic string) iter.Seq[peer.ID]
```
We collect the iter to a slice when using this. Maybe just change this to return a slice?

MarcoPolo replied:
the slices collect was for debugging. I can remove that.

sukunrt commented on partialmessages/partialmsgs.go:
```diff
@@ -0,0 +1,290 @@
+	}
+	e.Logger.Debug("Sending partial messages to peers", "peers", peers, "topic", topic, "group", groupID)
+	for _, p := range peers {
+		log := e.Logger.With("peer", p)
```
Will this allocate? It is fine as is, just curious.

MarcoPolo replied:
Yes, it does. Although I don't think it will stringify p until necessary.

sukunrt commented on partialmessages/partialmsgs.go:
```diff
@@ -0,0 +1,290 @@
+			// them what we can.
+			pm, rest, err := partial.PartialMessageBytes(peerState.partsMetadata)
+			if err != nil {
+				e.Logger.Warn("partial message extension failed to get partial message bytes", "error", err)
```
```suggestion
				log.Warn("partial message extension failed to get partial message bytes", "error", err)
```

sukunrt commented on partialmessages/partialmsgs.go:
```diff
@@ -0,0 +1,290 @@
+	ValidateRPC func(from peer.ID, rpc *pb.PartialMessagesExtension) error
+
+	// GroupTTLByHeatbeat is how many heartbeats we store Group state for.
+	GroupTTLByHeatbeat int
```
```suggestion
	// GroupTTLByHeatbeat is how many heartbeats we store Group state for after publishing a partial message for the group.
	GroupTTLByHeatbeat int
```

sukunrt commented on partialmessages/partialmsgs.go:
```diff
@@ -0,0 +1,290 @@
+			peerState = &partialMessagePeerState{}
+			state.peerState[from] = peerState
+		}
+		peerState.partsMetadata = rpc.PartsMetadata
```
This naming is confusing. Can we make it either `HasMetadata` or `WantsMetadata`

sukunrt replied:
Note: The state we store is what the peer *HAS* but some of the code reads like this is what the peer *WANTS*

MarcoPolo replied:
The code here assigns no meaning to this. The application decides if this is has or wants.

Open to a better name, but has/wantsMetadata isn't right.

sukunrt commented on partialmessages/partialmsgs.go:
```diff
@@ -0,0 +1,290 @@
+			peerState = &partialMessagePeerState{}
+			state.peerState[from] = peerState
+		}
+		peerState.partsMetadata = rpc.PartsMetadata
```
Maybe we should let the application set this. 

This update looks like a timing issue. 

Say: We have parts 1, 2, 5, 6 
1. Peer sends a message saying I have part 1, 2
2. We send parts 5, 6 and update the peerstate in `PublishMessage` to (1, 2, 5, 6)
3. Concurrently with 2, we receive message from the peer saying I now have parts (1, 2, 3, 4)

Now the new state may override the previous state. This will result in duplicate sending of parts 5, 6 on next publishPartial if we have any new part to send.

sukunrt replied:
Another way I can think of is to have another field for the peer which says 
`sentParts` and change this 
```
PartialMessageBytes(metadata []byte) ([]byte, []byte, error)
```
to
```
PartialMessageBytes(peerHasMetadata []byte, weSentMetadata []byte) ([]byte, newWeSentMetadata []byte, error)
```

We can then remove the second return value and just set it to `myPartsMeta`.

MarcoPolo replied:
maybe we need a application provided `mergePartsMetadata`?

Right now the merge function when we receive a partsMetadata from the peer is a last write wins. But in your example a `binary or` would be more appropriate.

We could also replace

```
PartialMessageBytes(metadata []byte) ([]byte, []byte, error)
```

With
```
PartialMessageBytes(metadata []byte) ([]byte, error)
```
And update the peers partsMetatdata with `merge(peerParts, ourParts)`

MarcoPolo replied:
done in: eb2f04a3205d

MarcoPolo committed:
partialmessages: add basic bitmap package

The bitmap is probably the common case, so add library support for it.

MarcoPolo committed:
partialmessages: add explicit MergePartsMetadata function

This allows a peer's local partsMetadata to merge with their remote
partsMetadata. As long as the underlying PartsMetadata type supports
CRDT-like addition, we can avoid sending the peer duplicates.

Take the following example (credit Sukun):

> Say: We have parts 1, 2, 5, 6
> 
> 1. Peer sends a message saying I have part 1, 2
> 2. We send parts 5, 6 and update the peerstate in PublishMessage to (1, 2, 5, 6)
> 3. Concurrently with 2, we receive message from the peer saying I now have parts (1, 2, 3, 4)

With this change we can now correctly store the fact that this peer
should have parts 1,2,3,4,5,6. And in a way that does not assume a
specific representation of the PartsMetadata.

sukunrt commented on gossipsub.go:
```diff
@@ -1375,6 +1375,10 @@ func (gs *GossipSubRouter) rpcs(msg *Message) iter.Seq2[peer.ID, *RPC] {
+			if peerStates, ok := gs.p.topics[topic]; ok && peerStates[pid].requestsPartial {
+				// The peer requested partial messages. We'll skip sending them full messages
+				continue
+			}
```
Do we need to check whether we support partialMessages?
Overall, can you add some documentation regarding this?

1. What happens when they support and we don't.
1. What happens when we support and they don't.

MarcoPolo replied:
done in 237d4182

sukunrt commented on gossipsub.go:
```diff
@@ -1375,6 +1375,10 @@ func (gs *GossipSubRouter) rpcs(msg *Message) iter.Seq2[peer.ID, *RPC] {
 			if pid == from || pid == peer.ID(msg.GetFrom()) {
 				continue
 			}
+			if peerStates, ok := gs.p.topics[topic]; ok && peerStates[pid].requestsPartial {
```
On line 871, there's IDONTWANT logic for incoming messages. We should skip sending IDONTWANT for partial message peers.

sukunrt commented on pubsub.go:
```diff
@@ -1192,9 +1241,16 @@ func (p *PubSub) doAnnounceRetry(pid peer.ID, topic string, sub bool) {
+		Partial:   &requestPartialMessages,
 	}
 
 	out := rpcWithSubs(subopt)
```
This needs to be sent with our extensions / Hello packet. Otherwise the first RPC we send the peer won't have the partial messages extension.

sukunrt replied:
fixed in: fe1f68da7d0119bdc0c568cd2e1c9307682a4087

I could see the impact in: https://github.com/libp2p/test-plans/pull/691

sukunrt replied:
Here's an alternative implementation: a638f7f040960b48975ca0c401552e7d701eff6d

MarcoPolo replied:
nice! I merged a slight variant of https://github.com/libp2p/go-libp2p-pubsub/commit/fe1f68da7d0119bdc0c568cd2e1c9307682a4087

sukunrt committed:
ensure that the Hello Packet is the first rpc sent

When a topic join happens concurrently with adding a new peer, we may
send the topic join announcement before the hello packet which contains
our extensions. This is a protocol violation.

This change starts sending rpcs on new stream only after we've enqued a
hello packet. We use UrgentPush to ensure that it's the first one sent.

sukunrt committed:
set write deadline for outgoing messages

MarcoPolo committed:
don't send IDONTWANT to partial message peers

if they requested partial messages on a given topic.

MarcoPolo committed:
add todo

MarcoPolo committed:
refactor score methods to accept topic string

instead of requiring a message object

MarcoPolo committed:
Add ability to update peer scores if using partial messages

We need some way to give peer feedback to gossipsub if the application
is using partial messages. This adds a new PeerFeedback method that lets
an application modify the peer score.

MarcoPolo committed:
nit: rename method from old form

MarcoPolo committed:
limit the number of peer initiated group states we track

MarcoPolo committed:
support fanout topics for partial messages

sukunrt reviewed (approved):
LGTM :tada:

MarcoPolo committed:
Add support for `supportsPartial`

This lets us distinguish who can support a partial message request
versus who requests partial message on a per topic basis.

MarcoPolo committed:
pb: add `supportsPartial` field in SubOpts

MarcoPolo committed:
add test for SupportsPartial subscribe option

Creates a mixed network where only one peer request partial, and the
rest either support or don't

MarcoPolo committed:
partialmessages: add per peer bounds on peer initiated groups

A peer initiated group state is a state that is initiatlized due to
receiving a partial message RPC for which we have not (yet) published
to.

It's useful to have this state so we can remember what parts a peer has
if we do publish to that group.

Previously we limited the total number of peer initiated group states in a
topic. This was agnostic to how many groups a single peer initiated,
which meant that a single peer could exhaust the whole limit.

This change tracks how many group states each peer has initialized and
limits each peer to a configurable number of these states. The default
total and per peer limits of 255 8 are likely good enough for most use
cases, but the options exist for users to change this if they need to.

MarcoPolo committed:
partialmessages: rename field to reflect use

changes partialMessageStatePerTopicGroup to
partialMessageStatePerGroupPerTopic to better reflect its usage.

MarcoPolo committed:
partialmessages: Add pairwise interaction test
